pinot-core/src/main/java/org/apache/pinot/core/operator/CombineGroupByOrderByOperator.java

```diff
@@ -0,0 +1,249 @@
+   * result block.
+   * <ul>
+   *   <li>
+   *     Concurrently merge group-by results from multiple result blocks into {@link org.apache.pinot.core.data.table.IndexedTable}
+   *   </li>
+   *   <li>
+   *     Set all exceptions encountered during execution into the merged result block
+   *   </li>
+   * </ul>
+   */
+  @Override
+  protected IntermediateResultsBlock getNextBlock() {
+    int numOperators = _operators.size();
+    CountDownLatch operatorLatch = new CountDownLatch(numOperators);
+
+    int numAggregationFunctions = _brokerRequest.getAggregationsInfoSize();
+    int numGroupBy = _brokerRequest.getGroupBy().getExpressionsSize();
+    ConcurrentLinkedQueue<ProcessingException> mergedProcessingExceptions = new ConcurrentLinkedQueue<>();
+
+    Future[] futures = new Future[numOperators];
+    for (int i = 0; i < numOperators; i++) {
+      int index = i;
+      futures[i] = _executorService.submit(new TraceRunnable() {
+        @SuppressWarnings("unchecked")
+        @Override
+        public void runJob() {
+          AggregationGroupByResult aggregationGroupByResult;
+
+          try {
+            IntermediateResultsBlock intermediateResultsBlock =
+                (IntermediateResultsBlock) _operators.get(index).nextBlock();
+
+            _initLock.lock();
+            try {
+              if (_dataSchema == null) {
+                _dataSchema = intermediateResultsBlock.getDataSchema();
+                _indexedTable.init(_dataSchema, _brokerRequest.getAggregationsInfo(), _brokerRequest.getOrderBy(),
+                    _interSegmentNumGroupsLimit, false);
+              }
+            } finally {
+              _initLock.unlock();
+            }
+
+            // Merge processing exceptions.
+            List<ProcessingException> processingExceptionsToMerge = intermediateResultsBlock.getProcessingExceptions();
+            if (processingExceptionsToMerge != null) {
+              mergedProcessingExceptions.addAll(processingExceptionsToMerge);
+            }
+
+            // Merge aggregation group-by result.
+            aggregationGroupByResult = intermediateResultsBlock.getAggregationGroupByResult();
+            if (aggregationGroupByResult != null) {
+              // Iterate over the group-by keys, for each key, update the group-by result in the indexedTable.
+              Iterator<GroupKeyGenerator.GroupKey> groupKeyIterator = aggregationGroupByResult.getGroupKeyIterator();
+              while (groupKeyIterator.hasNext()) {
+                GroupKeyGenerator.GroupKey groupKey = groupKeyIterator.next();
+                String[] stringKey = groupKey._stringKey.split(GroupKeyGenerator.DELIMITER);
+                Object[] objectKey = new Object[numGroupBy];
+                for (int i = 0; i < stringKey.length; i++) {
+                  DataSchema.ColumnDataType columnDataType = _dataSchema.getColumnDataType(i);
+                  switch (columnDataType) {
```

Comment:
is it possible to avoid the switch statement for each row? looks like the dataSchema remains the same within the loop. we can probably have an interface to convert string to Object and have the right implementations set in advance?

Reply:
Fixed in CombineGroupByOperator and BrokerReduce service, please take a look. 
However, I'm not able to achieve this in IntermediateResultsBlock, because that one is conversion from Object -> primitive data types. Any suggestions?

Reply:
thanks. will look into that.